File: compatibility-tests/src/test/java/nnl/rocks/kactoos/func/RetryFuncTest.java
/**
 * The MIT License (MIT)
 * <p>
 * Copyright (c) 2017-2018 Yegor Bugayenko
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package nnl.rocks.kactoos.func;

import org.hamcrest.MatcherAssert;
import org.hamcrest.Matchers;
import org.junit.Test;

import java.security.SecureRandom;

/**
 * Test case for {@link RetryFunc}.
 *
 * @author Yegor Bugayenko (yegor256@gmail.com)
 * @version $Id: 891f51ad523f379feeb3e383b5f4562172a84fae $
 * @since 0.8
 * @checkstyle JavadocMethodCheck (500 lines)
 */
public final class RetryFuncTest {

    @Test
    public void runsFuncMultipleTimes() throws Exception {
        MatcherAssert.assertThat(
            new RetryFunc<>(
                input -> {
                    // @checkstyle MagicNumberCheck (1 line)
                    if (new SecureRandom().nextDouble() > 0.3d) {
                        throw new IllegalArgumentException("May happen");
                    }
                    return 0;
                },
                Integer.MAX_VALUE
            ).apply(true),
            Matchers.equalTo(0)
        );
    }

}

File: kactoos-jvm/src/test/kotlin/nnl/rocks/kactoos/func/RetryFuncTest.kt

package nnl.rocks.kactoos.func

import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.Test

import java.security.SecureRandom

/**
 * Test case for [RetryFunc].
 *
 * @author Yegor Bugayenko (yegor256@gmail.com)
 * @version $Id: 891f51ad523f379feeb3e383b5f4562172a84fae $
 * @since 0.8
 * @checkstyle JavadocMethodCheck (500 lines)
 */
class RetryFuncTest {

    @Test
    @Throws(Exception::class)
    fun runsFuncMultipleTimes() {
        MatcherAssert.assertThat(
                RetryFunc<Any, Int>(
                        { input ->
                            // @checkstyle MagicNumberCheck (1 line)
                            if (SecureRandom().nextDouble() > 0.3) {
                                throw IllegalArgumentException("May happen")
                            }
                            0
                        },
                        Integer.MAX_VALUE
                ).apply(true),
                Matchers.equalTo(0)
        )
    }
}
